src/intersection-observer-polyfill.js

```diff
@@ -375,16 +404,33 @@ export class IntersectionObserverPolyfill {
     changeEntry.target = element;
     return changeEntry;
   }
+
+  /**
+   * Handle Mutation Oberserver events
+   * @private
+   */
+  handleMutationObserverNotification_() {
+    this.tick(this.viewport_.getRect());
```

Comment:
This is gonna fire a lot. `FixedLayer` is getting by with [checking](https://github.com/ampproject/amphtml/blob/master/src/service/fixed-layer.js#L208-L215) if a hidden attribute is changed, not _any_ attribute.

Reply:
visibility can be toggled by attributes other than `hidden`, for example an arbitrary attribute that has CSS selector with `display: none` on it.

Reply:
In my testing, this was only called on page load, and after all bind changes had resolved. But I'll go ahead, and add a throttle 😄

Reply:
@torch2424 from your experiment we see all mutation events are already batched right?
@jridgewell do you see otherwise?

Reply:
> for example an arbitrary attribute that has CSS selector with display: none on it.

Yes, but `FixedLayer`, [on-actions](https://www.ampproject.org/docs/interaction_dynamic/amp-actions-and-events#globally-defined-events-and-actions), and [amp-bind](https://www.ampproject.org/docs/reference/components/amp-bind#bindings) are taking the position that the only official way is the `hidden` attribute.

> this was only called on page load, and after all bind changes had resolved

This will be fired after every animation frame that has an attribute mutation. Like very amp-element's `buildCallback` and `layoutCallback`, etc.

> we see all mutation events are already batched right?

Correct, they're batched per animation frame.

Reply:
> Yes, but FixedLayer, on-actions, and amp-bind are taking the position that the only official way is the hidden attribute.

how about `attempCollapse` or even `attempChangeHeight`?

Reply:
`attemptCollapse` toggles `hidden` (so does expanding). `attemptChangeHeight` does not change the hidden visibility of an element.

Reply:
`attemptChangeHeight` could have changed sibling's height hence moved the target element in/out viewport.

Reply:
Correct, and there are numerous ways for CSS sibling selectors to change hidden visibility, too. The runtime's position is that only the `hidden` attribute is the proper way.

Reply:
not even CSS sibling selector (which might not be so common). amp-bind can dynamically change CSS classes on an element and toggle visibility. That's to say, if Fandango shopping work flow later has moved away from hidden attribute to something else, then the analytics bug would come again.

overall it's a tradeoff between visibility tracking accuracy / coverage VS runtime performance. I will let @cramforce make the call here.

Reply:
Is it worth creating a nasty "DisplayObserver polyfill" like this one?
https://output.jsbin.com/texepexaxa/5/quiet

It costs 1 iframe :)